### ISSUE TITLE
Printing a table

### ISSUE THREAD

**Issue by mansona** [offirgolan/ember-light-table#471]
Howdy folks 👋 

I'm wondering if anyone has had any success in trying to print an ember-light-table. I've tried a few things but because of some of the magic you are doing with height and two different tables (one for head and one for body) it seems impossible 🤔 

I would love to have something that I could pass into the rendering of the table to essentially just say "stop all magic and just give me a table", that way I could turn off the magic before printing. 

I am able to do this in Javascript because essentially my "printing" is happening in fastboot. 

Any thoughts? In the mean time, I have reverted to just building a "normal" table in my template 😂

**Comment by mansona** [comment]
I should have mentioned that the main issue with the table is that it is getting cut off at exactly one screen height because I have set the height to be `100vh`

**Comment by buschtoens** [comment]
I think the solution to this problem is two-fold:

1. #468: generate traditional "single `<table>`" markup in Fastboot
2. disable fixed headers for `@media print`

The latter one is already possible by overwriting our and `ember-scrollable`'s styles. However, it won't work for tables that have occlusion rendering (#435) enabled .

Ideally we would ship some default styles that user can opt in to and would expand scrollable tables to their full height on printers.

**Comment by buschtoens** [comment]
> I would love to have something that I could pass into the rendering of the table to essentially just say "stop all magic and just give me a table", that way I could turn off the magic before printing.

When #468 lands we could also provide an escape hatch to tables rendered in the browser, however I believe that would be the wrong approach for enabling printing.

**Comment by buschtoens** [comment]
I played around with some styles and I think media queries are a viable solution here.

![image](https://user-images.githubusercontent.com/834636/28886330-62d2f0a6-77b9-11e7-9766-32ed07e2edcd.png)
![image](https://user-images.githubusercontent.com/834636/28886342-6c39cf2a-77b9-11e7-937b-95221da6b431.png)

**Comment by mansona** [comment]
I know there is plenty of work still to do but man am I glad it's not just a simple fix (i.e. i'm not crazy) 😂 

That media query stuff that you were doing, was it using a "fixed height" i.e. setting the pixel height?  I was only asking because that last entry you have in your example is cut off. Also is there something between Brando and Bridget? i.e. are you missing an entry?

One thing that was interesting when I just built a _standard_ table for my needs is that it seems (in chrome at least) the table headers are repeated on each of the pages that are printed

**Comment by buschtoens** [comment]
> I know there is plenty of work still to do but man am I glad it's not just a simple fix (i.e. i'm not crazy) 

Well you already could make your light tables more printer-friendly by tweaking the CSS a bit. :stuck_out_tongue_winking_eye: 

> That media query stuff that you were doing, was it using a "fixed height" i.e. setting the pixel height? I was only asking because that last entry you have in your example is cut off. Also is there something between Brando and Bridget? i.e. are you missing an entry?

It's the example from our demo page. I manually disabled the fixed height (as a media query would do) and tweaked some other layout settings, in order for the table body to expand as tall as its content. 

There's no row missing between Brando and Bridgette.

The last row is indeed cut off, but I believe that's just some minor issue with margins and paddings that is not unfixable.

So to sum it up, no there's no row missing. :slightly_smiling_face: 

> One thing that was interesting when I just built a standard table for my needs is that it seems (in chrome at least) the table headers are repeated on each of the pages that are printed

That is standard behavior when using a single `<table>´. Ideally we would also enable this, but methinks that's really only possible when we have a single `<table>` element...  
We *could* possibly simulate a single `<table>` by rendering a second `<thead>` in the main `<table>` that's only visible for `@media print`.

I'd like to hide that behind a flag though, since it will incur a performance hit.

**Comment by Techn1x** [comment]
I managed to get some pretty nice looking printable tables, all I really did was disable the fixed header in my table, and remove the height declaration from the table (causing it to height auto). I have hundreds/thousands of records in my table and they definitely all appear in the print page
`{{#light-table table responsive=true as |t|}}`
`{{t.head fixed=false}}`

I also had to make some specific style changes in my print.scss but those were all changes specific to my website (removing website navbar, borders, setting body { overflow: visible !important; }  - that one caused my table to get cut off as well but is only an issue if you have your body element normally scrollable)

<img width="559" alt="untitled" src="https://user-images.githubusercontent.com/1049837/32252686-3e6c2ece-bee2-11e7-93a4-72911f0d6aa2.png">

**Comment by Techn1x** [comment]
Actually I might have spoken too soon. When I Save To PDF, and set the page to Landscape, some rows are split between pages (half of the row is printed on the page before).

I also noticed that (no matter what print options) at the end of each page an additional horizontal line is drawn, making it look like there is an empty row?? You can see this in the picture I attached above, right at the bottom of the first page.

**Comment by buschtoens** [comment]
@Techn1x [`page-break`](https://css-tricks.com/almanac/properties/p/page-break/) might be what you're looking for.

**Comment by Techn1x** [comment]
My webapp needs to print multi-page tables and I've come across some quirks when printing these large tables (mostly IE/Edge). In case this helps anyone, here's the SCSS I use to make large ELT's printer-friendly

```sass
.ember-light-table {
  // Fix #2: After applying fix #1, sometimes this class gets overflow, and it is set to overflow:auto, making a scrollbar appear for various browsers.
  // Either visible/hidden is fine, just not auto
  // overflow: visible !important;
  overflow: hidden !important;

  // Fix #3: After applying fix #1, Edge & IE will render all rows correctly, but IE & Firefox have a large gap after tables, due to flexbox bugs?
  display: block !important;

  .lt-body-wrap {
    // Fix #1: This prevents IE11/Edge issue for large tables where after page break, the rest of the table will not render.
    overflow: visible !important;

    // Fix #3: see above
    display: block !important;

    .lt-inline.lt-head {
      break-after: avoid; // Stop the header from being separated from the body of the table
      break-inside: avoid; // Stop the header from being broken (eg if using subcolumns)

      // Fix #4: On Chrome, if there is enough room at the end of the page for some of the header block, and a page break occurs inside it (despite the above),
      // it can cause the header to be rendered twice in a glitchy way (once at the end of the previous page, and again on the next page)
      // https://github.com/offirgolan/ember-light-table/issues/579
      // Need to set the header block to not be 'table-header-group', which seems to cause this behaviour
      thead.lt-head {
        display: table-footer-group;
      }
    }
  }

  // Don't allow page breaks inside cells
  // NOTE: th's are basically the same as td
  table tr, table tr td, table tr th {
    break-inside: avoid;
  }
}
```

Used with basic ELT markup like
```handlebars
{{#light-table table
  responsive=false
  as |t|}}
  {{t.head}}
  {{t.body}}
{{/light-table}}
```
